File: src/Finos.Fdc3.Backplane.Client/Broadcast/Channels/ChannelClient.cs
/**
	* SPDX-License-Identifier: Apache-2.0
	* Copyright 2021 FINOS FDC3 contributors - see NOTICE file
	*/

using Finos.Fdc3.Backplane.Client.Middleware;
using Finos.Fdc3.Backplane.DTO.FDC3;
using System;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

[assembly: InternalsVisibleTo("Finos.Fdc3.Backplane.Client.Test")]
namespace Finos.Fdc3.Backplane.Client.Broadcast.Channels
{
    internal class ChannelClient : IChannel
    {
        private readonly IClientMiddleware _clientMiddleware;

        public ChannelClient(IClientMiddleware clientMiddleware, Channel channelDto)
        {
            _clientMiddleware = clientMiddleware;
            Id = channelDto.Id;
            Type = channelDto.Type;
            DisplayMetadata = channelDto.DisplayMetadata;
        }

        public string Id { get; }

        public string Type { get; }

        public DisplayMetadata DisplayMetadata { get; }

        public async Task<IListener> AddContextListenerAsync(string contextType, Action<IContext> handler, CancellationToken ct = default)
        {
            return await _clientMiddleware.AddContextListenerAsync(contextType, handler, Id, ct);
        }

        public async Task BroadcastAsync(IContext context, CancellationToken ct = default)
        {
            await _clientMiddleware.BroadcastAsync(context, Id, ct);
        }

        public async Task<IContext> GetCurrentContextAsync(string contextype = null, CancellationToken ct = default)
        {
            IContext context = await _clientMiddleware.GetCurrentContextAsync(Id, ct);
            if (contextype == null || contextype == context?.Type)
            {
                return context;
            }
            return null;

        }

        public override string ToString()
        {
            return $"{Id}-{Type}-{DisplayMetadata?.Name}";
        }
    }
}

File: src/Finos.Fdc3.Backplane.Client.JS/src/broadcast/channels/ChannelClient.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2021 FINOS FDC3 contributors - see NOTICE file
 */

import { Channel, Context, ContextHandler, DisplayMetadata, Listener } from '@finos/fdc3';
import { ClientMiddleware } from '../../service/ClientMiddleware';
export class Fdc3Channel implements Channel {
	id: string;
	type: string;
	displayMetadata?: DisplayMetadata | undefined;
	private clientMiddleware: ClientMiddleware;
	private currentContext: Context | null = null;

	constructor(id: string, type: string, desktopAgentService: ClientMiddleware, displayMetadata?: DisplayMetadata) {
		this.id = id;
		this.type = type;
		this.clientMiddleware = desktopAgentService;
		this.displayMetadata = displayMetadata;
	}

	broadcast(context: Context): void {
		this.currentContext = context;
		this.clientMiddleware.broadcastContext(context, this.id);
	}

	getCurrentContext(contextType?: string): Promise<Context | null> {
		return Promise.resolve(this.currentContext);
	}

	addContextListener(handler: ContextHandler): Listener;
	addContextListener(contextType: string | null, handler: ContextHandler): Listener;
	addContextListener(contextType: any, handler?: any): Listener {
		let contextListener: any = null;
		contextListener = this.clientMiddleware.addContextListener(contextType, handler, this.id);
		return contextListener;
	}
}
